src/dotnet/ToolPackage/ToolPackageInstance.cs

```diff
@@ -0,0 +1,172 @@
+using System;
+using System.IO;
+using System.Linq;
+using System.Collections.Generic;
+using Microsoft.DotNet.Cli;
+using Microsoft.DotNet.Tools;
+using Microsoft.Extensions.EnvironmentAbstractions;
+using NuGet.ProjectModel;
+
+namespace Microsoft.DotNet.ToolPackage
+{
+    // This is named "ToolPackageInstance" because "ToolPackage" would conflict with the namespace
+    internal class ToolPackageInstance : IToolPackage
+    {
+        private Lazy<IReadOnlyList<ToolCommand>> _commands;
+
+        public ToolPackageInstance(
+            string packageId,
+            string packageVersion,
+            DirectoryPath packageDirectory)
+        {
+            if (packageId == null)
+            {
+                throw new ArgumentNullException(nameof(packageId));
+            }
+            if (packageVersion == null)
+            {
+                throw new ArgumentNullException(nameof(packageVersion));
+            }
+
+            PackageId = packageId;
+            PackageVersion = packageVersion;
+            PackageDirectory = packageDirectory;
+            _commands = new Lazy<IReadOnlyList<ToolCommand>>(GetCommands);
+        }
+
+        public string PackageId { get; private set; }
+
+        public string PackageVersion { get; private set; }
+
+        public DirectoryPath PackageDirectory { get; private set; }
+
+        public IReadOnlyList<ToolCommand> Commands
+        {
+            get
+            {
+                return _commands.Value;
+            }
+        }
+
+        public void Uninstall()
```

Comment:
it is odd that install and uninstall are in different level. but it is not a big deal

Reply:
I feel it is better to move out. What uninstall really need is folder structure logic and packageid package version. But since there is only few code. It is fine to keep it there. If it ever get bigger, we should create a new class.

Reply:
👍